presto-druid/src/main/java/com/facebook/presto/druid/DruidConnectorPlanOptimizer.java

```diff
@@ -0,0 +1,229 @@
+                .orElseThrow(() -> new PrestoException(GENERIC_INTERNAL_ERROR,
+                        "Expected to find druid table handle for the scan node"));
+        return maxSubplan.accept(new Visitor(druidTableScanNode, session, idAllocator), null);
+    }
+
+    private static Optional<DruidTableHandle> getDruidTableHandle(TableScanNode tableScanNode)
+    {
+        TableHandle table = tableScanNode.getTable();
+        if (table != null) {
+            ConnectorTableHandle connectorHandle = table.getConnectorHandle();
+            if (connectorHandle instanceof DruidTableHandle) {
+                return Optional.of((DruidTableHandle) connectorHandle);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<TableScanNode> getOnlyDruidTable(Map<TableScanNode, Void> scanNodes)
+    {
+        if (scanNodes.size() == 1) {
+            TableScanNode tableScanNode = scanNodes.keySet().iterator().next();
+            if (getDruidTableHandle(tableScanNode).isPresent()) {
+                return Optional.of(tableScanNode);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static PlanNode replaceChildren(PlanNode node, List<PlanNode> children)
+    {
+        for (int i = 0; i < node.getSources().size(); i++) {
+            if (children.get(i) != node.getSources().get(i)) {
+                return node.replaceChildren(children);
+            }
+        }
+        return node;
+    }
+
+    private static class TableFindingVisitor
+            extends PlanVisitor<Map<TableScanNode, Void>, Void>
+    {
+        @Override
+        public Map<TableScanNode, Void> visitPlan(PlanNode node, Void context)
+        {
+            Map<TableScanNode, Void> ret = new IdentityHashMap<>();
+            node.getSources().forEach(source -> ret.putAll(source.accept(this, context)));
+            return ret;
+        }
+
+        @Override
+        public Map<TableScanNode, Void> visitTableScan(TableScanNode node, Void context)
+        {
+            Map<TableScanNode, Void> ret = new IdentityHashMap<>();
+            ret.put(node, null);
+            return ret;
+        }
+    }
+
+    private class Visitor
+            extends PlanVisitor<PlanNode, Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final ConnectorSession session;
+        private final TableScanNode tableScanNode;
+        private final IdentityHashMap<FilterNode, Void> filtersSplitUp = new IdentityHashMap<>();
+
+        public Visitor(TableScanNode tableScanNode, ConnectorSession session, PlanNodeIdAllocator idAllocator)
+        {
+            this.session = session;
+            this.idAllocator = idAllocator;
+            this.tableScanNode = tableScanNode;
+            // Just making sure that the table exists
+            getDruidTableHandle(this.tableScanNode).get().getTableName();
+        }
+
+        private Optional<PlanNode> tryCreatingNewScanNode(PlanNode plan)
```

Comment:
I wonder why this code looks so familiar :-)

Reply:
haha, yes, I copied a bunch of code from ur Pinot connector :)

Reply:
Hi @zhenxiao , It would help the reader if you could refer to the PR or commit of the Pinot connector optimizer PR/commit in your commit message. This code is basically identical to the Pinot connector and it would not only help with proper attribution but also if we subsequently need to fix some bugs in both the pinot/druid connectors. 

Also, it would be great if you could think a bit about how we may refactor to share more common code b/w the pinot and the druid connector. What did you think about what we could refactor into a separate module ? 

Thanks

Reply:
Good idea. Yep, Pinot and Druid are similar. We could think about refactoring common code. I will add "Reference Pinot Connector" in the commit message. The optimizer code is similar. The code are not identical, Druid has its own segment scan log and broker page source. Logic for `FilterExpressionConverter` are different, too.